### ISSUE TITLE
--from-to provides insufficent failure information

### ISSUE THREAD

**Issue by cdiskan** [Azure/azure-storage-azcopy#2153]
##### Note: The version is visible when running AzCopy without any argument

AzCopy 10.18.0

### Which platform are you using? (ex: Windows, Mac, Linux)

Windows

##### Note: Please remove the SAS to avoid exposing your credentials. If you cannot remember the exact command, please retrieve it from the beginning of the log file.

.\azcopy.exe copy $ChildSource $ChildDestSASstring --recursive --overwrite=false  --preserve-smb-permissions  --preserve-smb-info --from-to localfiles

### What problem was encountered?

When using the --from-to switch, the failure information provides insufficient information on acceptable options. 

The error provided is "fatal: from-to argument required, PipeBlob (upload) or BlobPipe (download) is acceptable" . However, there are numerous from-to option pairs available. 

From reading the .go file at https://github.com/Azure/azure-storage-azcopy/blob/main/cmd/copy.go the following options appear to be defined:

BenchmarkBlob
BenchmarkBlobFS
BenchmarkFile
BlobBlob
BlobFile
BlobFSLocal
BlobFSNone
BlobFSTrash
BlobLocal
BlobNone
BlobPipe
BlobTrash
FileBlob
FileFile
FileLocal
FileNone
FileTrash
GCPBlob
LocalBlob
LocalBlobFS
LocalFile
PipeBlob
S3Blob

### How can we reproduce the problem in the simplest way?

Use AZCopy with the --from-to switch, with any invalid from-to response.

### Have you found a mitigation/solution?

No workaround available as the issue is with an insufficient error, not an application function.

**Comment by adreed-msft** [comment]
Added @normesta to help brainstorm good ways to explain the `--from-to` flag without getting too wordy or just listing every option out as it may be overwhelming to look at.

I was thinking something along the lines of `from-to indicates a SourceDestination pairing (e.g. LocalBlob). Valid resource names are Local, Blob, File, Trash, S3, GCP`

**Comment by normesta** [comment]
Hi @adreed-msft  - thanks for the ping! Because these values appear no where else and the ref is the ultimate source of truth, I vote to include all of them. A comma separated list seems fine. Most of these combinations seem fairly intuitive, so I'd list them and see if any customers still express confusion. Also (in my opinion), there's no requirement to keep the explanations of parameters short. The right length expresses all the information needed to understand the parameters use while reducing extraneous information where possible. I think the parameter definition should be flushed out a bit more. I might suggest something like this: 

"Specifies the type of object to copy from the source location and the type of object that you want to exist at the destination when the operation completes.  Use this parameter if the source argument of the copy command is not a storage service URL because without this parameter, AzCopy won't be able to infer the correct object type. Valid combinations include the following: <comma separated list>"

**Comment by cdiskan** [comment]
@adreed-msft 

I would second @normesta , in that a detailed error would likely be the most helpful for customers seeking to understand and utilize the tool. The current documentation of the AZCopy switches is still sparse in many places, and a detailed error message helps ensure that users can self-resolve issues quickly without needing to go search for docs\man pages. 

I would encourage someone who has a deeper understanding of the --from-to switch to review my list, as it was unclear if some of the from\to options (Such as "Benchmark" or "None") are actually functional from a customer perspective, or if they are restricted to other internal application functions. 

Thanks!

**Comment by normesta** [comment]
@adreed-msft Also, if it's helpful for heading off future support cases, I'd be happy to add a backlog item for myself in a future sprint to go through and add more detailed content to the ref files inline. If we believe there's impact in that task, I'd much rather update the content in source and import to ref vs. update our ref docs. That way the two (inline and doc ref) remain in sync and there's always one source of truth (the inline ref). When using search, I do think a lot of folks find the ref articles much more often than they discover our cleaner conceptual articles. Instead of hoping folks will find the conceptual content or trying to game search, it might make sense to invest more deeply in just making the ref more detailed and complete. Open to ideas and glad to help there as AzCopy is a very popular tool.

**Comment by nakulkar-msft** [comment]
The bug here is that we are incorrectly parsing source and destination combo. from-to should not be necessary until the user is piping.

**Comment by wilsonlmh** [comment]
I don't find any document mentioning what is Pipe, Local, Blob, File, Trash, S3, and GCP, at least not public document.
What is the purpose of list something out without explanation?
This very example demonstrated a "hidden" and maybe only effective way to delete a folder recursively:
https://github.com/Azure/azure-storage-azcopy/issues/1294#issuecomment-2788564375
I guess it's easier for all of us if some more explanation exists?
